src/EditorFeatures/Core/Implementation/NavigateTo/AbstractNavigateToSearchService.cs

```diff
@@ -1,134 +1,58 @@
-            //
-            // So the question becomes: what do we do if we have multiple match results, and we
-            // need to map to a single MatchKind.
-            //
-            // First, consider a main reason we have multiple match results.  And this happened
-            // when the user types a dotted name (like "Microsoft.CodeAnalysis.ISymbol").  Such
-            // a name would match actual entities: Microsoft.CodeAnalysis.ISymbol *and* 
-            // Microsoft.CodeAnalysis.IAliasSymbol.  The first will be an [Exact, Exact, Exact] 
-            // match, and the second will be an [Exact, Exact, CamelCase] match.  In this
-            // case our belief is that the names will go from least specific to most specific. 
-            // So, the left items may match lots of stuff, while the rightmost items will match
-            // a smaller set of items.  As such, we use the last pattern match to try to decide
-            // what type of editor MatchKind to map to.
-            if (containsDots)
+            else if (_resultProviders.Length == 1)
             {
-                var lastResult = matchResult.LastOrNullable();
-                if (lastResult.HasValue)
-                {
-                    switch (lastResult.Value.Kind)
-                    {
-                        case PatternMatchKind.Exact:
-                            return MatchKind.Exact;
-                        case PatternMatchKind.Prefix:
-                            return MatchKind.Prefix;
-                        case PatternMatchKind.Substring:
-                            return MatchKind.Substring;
-                    }
-                }
+                return await _resultProviders[0].SearchProjectAsync(project, searchPattern, cancellationToken).ConfigureAwait(false);
             }
             else
             {
-                // If it wasn't a dotted name, and we have multiple results, that's because they
-                // had a something like a space separated pattern.  In that case, there's no
-                // clear indication as to what is the most important part of the pattern.  So 
-                // we make the result as good as any constituent part.
-                if (matchResult.Any(r => r.Kind == PatternMatchKind.Exact))
+                var tasks = new Task<IEnumerable<INavigateToSearchResult>>[_resultProviders.Length];
+                for (int i = 0; i < _resultProviders.Length; i++)
                 {
-                    return MatchKind.Exact;
-                }
+                    cancellationToken.ThrowIfCancellationRequested();
 
-                if (matchResult.Any(r => r.Kind == PatternMatchKind.Prefix))
-                {
-                    return MatchKind.Prefix;
+                    var resultProvider = _resultProviders[i];
+                    tasks[i] = Task.Run(() => resultProvider.SearchProjectAsync(project, searchPattern, cancellationToken), cancellationToken);
```

Comment:
Concurrent execution of providers is another open question: should we run them in parallel? should they run on background threads (we do so for analyzers)? I'll add this to the issue tracking the design.